### ISSUE TITLE
Provide wrappers around tf.layers

### ISSUE THREAD

**Issue by ppwwyyxx** [ppwwyyxx/tensorpack#291]
Since `tf.layers` is in tf core they'll hopefully be well maintained. It's easier to just call them instead of rewriting the layers.
Would also solve #289, #283, #236.
Before this feature is added, calling `tf.layers` directly would also work.

The only issue now is the name difference. Tensorpack has been using different argument names and variable names.

We could deprecate the old names, or add some global/layer-specific option to switch between the two.
Deprecation of variable names is probably not an option because models are hard to fix.

**Comment by PatWie** [comment]
Does `tf.layers` follows a common parameter pattern (name, input, param1, param2)? Does their BatchNorm provides "use_local_stats"?

I also thought about a "@tflayer_register()", which is probably the smoothest transition towards `tf.layers`. Hopefully, the output of "incoming-shape, outgoing shape, the number of params, size" will still exist in the future tensorpack versions.

**Comment by ppwwyyxx** [comment]
Yes I hope to keep the shape output. Just want to use tf.layers for the actual implementation.

I guess first step we can switch the implementation but keep all the APIs/names unchanged (by custom_getter).

**Comment by Skylion007** [comment]
We can also try autorefactor tools to update all the examples and put deprecation warning under the old layers. Python [rope](https://github.com/python-rope/rope/blob/master/docs/rope.rst) seems like the best solution if we do that approach.

**Comment by Skylion007** [comment]
Any update on this front? The longer we wait to fix this, the more troublesome an issue it will become.

**Comment by ppwwyyxx** [comment]
The change is supposed to be transparent to users. There won't be any trouble as said in the issue.

**Comment by Skylion007** [comment]
The question then begins if we create new layers should we support the Tensorflow arguments or the legacy Tensorpack ones? How can we easily mix the argscopes between Tensorpack and Tensorflow? Or is that what the custom getter is trying to achieve? And if Tensorpack layers are not being deprecated it becomes confusing when to use which one. Should the user register a new tf.layer or a Tensorpack layers? This seems like it is going to cause a lot of issues down the line if we don't eventually merge or deprecate the Tensorpack layers.

**Comment by ppwwyyxx** [comment]
You use whichever you like, and implement new layers in whichever ways you like. They are just doing the same thing with different interface (different function name, argument name, etc) and you can always use one in place of anther. Why would there be confusion?

**Comment by ppwwyyxx** [comment]
Closing. Only two layers were not switched to `tf.layers` implementation:
1. Conv2D. `tf.layers` doesn't support group conv.
2. BatchNorm. `tf.layers` doesn't support use_local_stat != is_training.
